Add assign kwarg to module.load_state_dict

mikaylagawarecki opened the pull request:
Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...


Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #102212

mikaylagawarecki committed:
Add assign kwarg to module.load_state_dict

[ghstack-poisoned]

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -2005,6 +2022,11 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
             exists in :attr:`state_dict`, :meth:`load_state_dict` will raise a
             ``RuntimeError``.
         """
+        if assign:
```
@jbschlosser the case you mentioned where we accidentally do 

```
with torch.device('meta'):
    m = Module()
opt = torch.optim.SGD(m.params())
m.load_state_dict(state_dict, assign=True)
out = net(inp)
out.sum().backward()
opt.step()
opt.zero_grad()
```

doesn't seem to error, it seems like [all the p.grad here](url) will be None so it would just be silently incorrect, so decided to warn here instead

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -2021,6 +2043,7 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
 
         def load(module, local_state_dict, prefix=''):
             local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
+            local_metadata['assign_to_params_buffers'] = assign
```
This might be a very debatable design choice here to put the flag into the `local_metadata` dict.

I chose to do this instead of adding a new `assign` argument to `_load_from_state_dict` because it seems like `_load_from_state_dict` is overridden **a lot** in practice. 

So if I had 
module._load_from_state_dict(
                local_state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs, assign)
here it would break with every module that overrides `_load_from_state_dict`. 

We could fix all the modules we own (and a ton more after ghimport) but it would still break a lot of external modules so I'm very hesitant to change the signature of `_load_from_state_dict`... 

Wondering whether there is a better way around this that I'm missing here?

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -2021,6 +2043,7 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
 
         def load(module, local_state_dict, prefix=''):
             local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
+            local_metadata['assign_to_params_buffers'] = int(assign)
```
This might be a very debatable design choice here to put the flag into the `local_metadata` dict.

I chose to do this instead of adding a new `assign` argument to `_load_from_state_dict` because it seems like `_load_from_state_dict` is overridden **a lot** in practice. 

So if I had 
```
module._load_from_state_dict(
                local_state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs, assign)
```
below it would break with every module that overrides `_load_from_state_dict`. 

We could fix all the modules we own (and a ton more after ghimport) but it would still break a lot of external modules so I'm very hesitant to change the signature of `_load_from_state_dict`... 

Wondering whether there is a better way around this that I'm missing here?

albanD replied:
That sounds like an ok workaround for me

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1954,7 +1959,15 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                            # Shape checks are already done above
+                            is_persistent_buffer = name in persistent_buffers
+                            if is_persistent_buffer:
+                                self.__setattr__(name, input_param)
```
nit: `setattr(self, name, input_param)` is the usual way to do this I think

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1896,6 +1896,10 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
         the version number at `local_metadata.get("version", None)`.
+        Additionally, :attr:`local_metadata` can also contain the key
+        `assign_to_params_buffers` that indicates whether keys that are meta
+        tensors in the module should be assigned their corresponding tensor
```
Not sure why there is a mention to meta Tensor here? They will all be assigned without distinction on the curr device right?

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1954,7 +1959,15 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                        if assign_to_params_buffers:
+                            # Shape checks are already done above
+                            is_persistent_buffer = name in persistent_buffers
+                            if is_persistent_buffer:
```
Why is this based on persistent buffer and not `if isinstance(param, torch.nn.Parameter)` ?

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1994,6 +2007,10 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
                 :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
+            assign (bool, optional): whether to assign items in the state
+            dictionary to their corresponding keys in the module instead
+            of copying them inplace into the module's parameters and buffers.
```
```suggestion
            of copying them inplace into the module's current parameters and buffers.
```

albanD commented on torch/nn/modules/module.py:
```diff
@@ -2021,6 +2043,7 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
 
         def load(module, local_state_dict, prefix=''):
             local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
+            local_metadata['assign_to_params_buffers'] = int(assign)
```
Why convert it to an int?

mikaylagawarecki replied:
Using a bool here makes [this legacy jit test ](https://www.internalfb.com/code/fbsource/[5d916f94b781450e74bda5e477213319be1948cd]/fbcode/caffe2/test/jit/test_module_apis.py?lines=38) fail with  

```
RuntimeError: Tracer cannot infer type of {'version': 1, 'assign_to_params_buffers': False}
2023-05-24T21:55:39.2362236Z :Dictionary inputs to traced functions must have consistent type. Found int and bool
``` :/

albanD replied:
cc @davidberard98 do you know what this error is about? Are you ok with us silency this legacy test?

mikaylagawarecki replied:
It occurred to me that we can make this backward compatible by not adding assign to the dict if it is False. I have just made this change torch.jit.script will probably fail if someone uses it with load_state_dict assign=True but I suppose that is unlikely (?)

davidberard98 replied:
fwiw, I think state dict is something that torchscript users care about so we shouldn't break it in the general case. But I think it's probably fine if assign=True doesn't work.

cc @IvanKobzarev who I think is working on state_dict torchscript stuff for torchrec purposes, to confirm that this isn't something that he cares about

albanD commented on test/test_nn.py:
```diff
@@ -2582,6 +2582,127 @@ def set_extra_state(self, state):
+        # Make sure parameters and persistent buffers were assigned
+        net_meta_state_dict = net_meta.state_dict()
+        for key in state_dict.keys():
+            self.assertTrue(state_dict[key].data_ptr() == net_meta_state_dict[key].data_ptr())
```
Can't this be `state_dict[key] is net_meta_state_dict[key]` if you pass `keep_vars=True` when getting the state_dicts?

mikaylagawarecki replied:
Hm, I guess because [here](https://github.com/pytorch/pytorch/blob/e6e9fd9e7eeda6668b26557f2ad07d4751af68ca/torch/nn/modules/module.py#L1968) during assignment I always wrap in `torch.nn.Parameter` (which could take either an `nn.Parameter` or a Tensor as input), they are actually different python objects

albanD replied:
Ho that is interesting. Commented below but I think it would be a great property for us to preserve during assignment.
Making new Tensor object is cheap and we do it easily. But when subclasses are involved. We don't know how expensive a new object will be.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1994,6 +2007,10 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
+            assign (bool, optional): whether to assign items in the state
+            dictionary to their corresponding keys in the module instead
+            of copying them inplace into the module's parameters and buffers.
+            Default: ``False``
```
I think here might be a good place to have a quick "Warning: if you use this option, you MUST create the optimizer after loading the state dict."

albanD commented on torch/nn/modules/module.py:
```diff
@@ -2005,6 +2022,11 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
             ``RuntimeError``.
         """
+        if assign:
+            warnings.warn("If using ``load_state_dict`` with ``assign=True``, "
```
From offline discussion, I think we should move this to Optimizer itself when a call to step() would do no actual work.

mikaylagawarecki replied:
Actually, looking more into how the optimizer is written, 

what happens is that for each param group, an optimizer will call `[init_group](https://github.com/pytorch/pytorch/blob/main/torch/optim/sgd.py#L73)` which updates the list of `params_with_grad` and then calls the optimizer step on that param group.

I wonder whether it might be a common use case of param_groups to freeze one param group and train another (i.e. some params_with_grad lists will be intentionally empty). If this is indeed a common use case and we check whether this list is empty and warn then, it might be somewhat noisy.

We could alternatively keep a separate list of all the params across all param groups and warn at the end if this list is empty, but I feel like this is somewhat wasteful.

Are we happy with maybe just warning in the documentation?

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

janeyx99 commented on test/optim/test_optim.py:
```diff
@@ -1607,6 +1607,42 @@ def closure():
+            inp = torch.randn(4, 5)
+            out = m(inp)
+            out.sum().backward()
+            with self.assertWarnsRegex(UserWarning, "None of the parameters"):
```
Could we get a more indicative snippet of the warning message here?

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -76,6 +76,27 @@ def _default_to_fused_or_foreach(params: List[torch.Tensor],
+def _warn_step_no_param_with_grad(stacklevel=2):
+
+    def is_first_time():
+        if not hasattr(_warn_step_no_param_with_grad, 'has_warned'):
+            return True
+        else:
+            return not _warn_step_no_param_with_grad.__dict__['has_warned']
```
```suggestion
        return not hasattr(_warn_step_no_param_with_grad, 'has_warned') or not _warn_step_no_param_with_grad.__dict__['has_warned']:
```
Could DeMorgan's Law it for further effect but this may be clearer.

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1954,7 +1958,14 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
-                        param.copy_(input_param)
+                        if assign_to_params_buffers:
+                            # Shape checks are already done above
+                            if isinstance(param, torch.nn.Parameter):
```
```suggestion
                            if isinstance(param, torch.nn.Parameter) and not isinstance(input_param, torch.nn.Parameter):
```

I think it would be great to be able to preserve `is` property here!

janeyx99 commented on torch/optim/adadelta.py:
```diff
@@ -118,6 +122,9 @@ def step(self, closure=None):
                 differentiable=differentiable,
             )
 
+        if not has_any_param_with_grad:
```
The logic here can be combined with lines 108-110, right? In general, I believe this entire logic could go together under init_group instead of here.

However, before you implement any of those changes, a concern I have is that we're adding this same exact code to every optimizer. This is fragile because it is easy to forget to keep this code consistent across optimizers or when we add new ones. A better solution would be to use `register_optimizer_step_pre_hook` and register a pre-hook for all optimizers in optimizer.py. The hook has access to the optimizer, so you should be able to access the params with optimizer.param_groups. This WOULD trade off the performance as it adds an extra loop over the params.

So could we benchmark the following solutions across diff configs of the number of params (say 10, 100, 1000, 10000, etc):
1. Group the logic here into init_group
2. Registering a hook into that same logic

mikaylagawarecki replied:
@janeyx99 


Ran the simple benchmarking script below and got the following results


**With global pre-hook**

```
def _global_warn_step_no_param_with_grad_pre_hook(opt: Optimizer, args: Tuple[Any], kwargs: Dict[Any, Any]):
    any_param_has_grad = any([p.grad is not None for param_group in opt.param_groups for p in param_group['params']])
    if not any_param_has_grad:
        _warn_step_no_param_with_grad()

register_optimizer_step_pre_hook(_global_warn_step_no_param_with_grad_pre_hook)
```
Mapping from num_layers to time
{1: 0.005815193653106689, 10: 0.015650224685668946, 100: 0.09916568994522094, 1000: 1.0547807788848877}

**With current method (no extra loop over params)**

{1: 0.005666229724884033, 10: 0.014990105628967285, 100: 0.09507855415344238, 1000: 1.0070961976051331}

Does it suffice to say that the difference is significant enough that we would want to add any_param_has_grad as an optimizer attribute that can be accessed by the hook so we don't have to do the double loop?

<details><summary><b>Click for script</b></summary>
```
import torch
import torch.nn as nn
import time

torch.manual_seed(5)
num_layers = [1, 10, 100, 1000]
times = dict()

for num_layers in num_layers:
    sum_overall_step_time = 0
    for i in range(100):
        print(i)
        model = torch.nn.Sequential(*(nn.Linear(64, 64) for _ in range(num_layers)))
        inp = torch.randn(64, 64)
        opt = torch.optim.SGD(model.parameters(), lr=0.001)

        num_iters = 100
        overall_step_time = 0
        for i in range(num_iters):
            opt.zero_grad()
            out = model(inp)
            out.sum().backward()
            before_step = time.time()
            opt.step()
            after_step = time.time()
            step_time = after_step - before_step
            overall_step_time += step_time
        sum_overall_step_time += overall_step_time

    average_overall_step_time = sum_overall_step_time / 100
    times[num_layers] = average_overall_step_time
print(times)
``` 

</details>

janeyx99 replied:
Yea

mikaylagawarecki replied:
I used a global post hook instead but I suppose there is still some duplicate code :/ what do you think?

mikaylagawarecki replied:
From offline discussion, perhaps the post hook isn't giving enough value in terms of code consistency here, since there is already a fair bit of duplicate code if we want to only loop over params once. Further, adding an attribute to optimizers so that the post hook can access it causes a lot of PT2 failures

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

janeyx99 commented on test/optim/test_optim.py:
```diff
@@ -1607,6 +1607,43 @@ def closure():
             self.assertEqual(old_param, param)
 
+    def test_step_warns_once_for_no_grads(self):
+        # not including LBFGS since it creates a zero-ed tensor in _gather_flat_grad
```
Wait..but should it? Is this warning no longer a concern with LBFGS? My impression is that this would require special casing for LBFGS, not that the warning should not happen. I also observe Rprop is missing.

janeyx99 commented:
Taking a step back, is the optimizer the earliest place to catch this warning? If the assign is misused, is there nothing we can check before the optimizer step that will notice something is wrong?

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

mikaylagawarecki commented:
From offline discussion -- we will not add the warning in the optimizers and just add a warning in the documentation for now.

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1982,18 +1994,26 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                 :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
+            assign (bool, optional): whether to assign items in the state
+                dictionary to their corresponding keys in the module instead
+                of copying them inplace into the module's current parameters and buffers.
```
nit: I would add one more sentence here similar to (feel free to rephrase): When False, the properties of the Tensors in the current Module are preserved while when True, the properties from the Tensors in the state dict are preserved.

albanD reviewed (approved):
Small nit on the doc but SGTM otherwise!

mikaylagawarecki committed:
Update on "Add assign kwarg to module.load_state_dict"

Fixes #64601 and #98906

Adds an `assign` argument to `load_state_dict` that loads params/buffers by assignment instead of doing `param.copy_(param_from_state_dict)`. 

Primarily intended to remove the need for the `.to_empty()` in 

```
with torch.device('meta'):
    m = SomeModule()
m.to_empty()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict)
```

so we can instead do 

```
with torch.device('meta'):
    m = SomeModule()
state_dict = torch.load('...pth')
m.load_state_dict(state_dict, assign=True)
```

**A problem with this PR for the case where the model is initialized on meta is what happens to nonpersistent buffers/params corresponding to keys missing from the state dict?**
What happens in the case where `load_state_dict(state_dict, strict=False, assign=True)` and the state_dict is missing some keys? The corresponding params missing from the `state_dict` and nonpersistent buffers would still be on `meta` and need to be manually initialized. However, I don't think we offer an API that would initialize these.

One solution would be to make these empty tensors but it might not be semantically correct...





[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge -f "distributed failures are unrelated"
